src/Controller/ViewerController.php

```diff
@@ -63,63 +63,42 @@ public function __construct(
    *   TRUE to open Collabora Online in edit mode.
    *   FALSE to open Collabora Online in readonly mode.
    *
+   * @throws \Symfony\Component\HttpKernel\Exception\BadRequestHttpException
+   *   The Collabora server or editor is not available, or mismatching scheme.
+   *
    * @return \Symfony\Component\HttpFoundation\Response
    *   Response suitable for iframe, without the usual page decorations.
    */
-  public function editor(MediaInterface $media, Request $request, $edit = FALSE) {
+  public function editor(MediaInterface $media, Request $request, $edit = FALSE): Response {
     try {
       $wopi_client_url = $this->discovery->getWopiClientURL();
     }
     catch (CollaboraNotAvailableException $e) {
-      $this->logger->warning(
-        "Collabora Online is not available.<br>\n" . Error::DEFAULT_ERROR_MESSAGE,
-        Error::decodeException($e) + [],
-      );
-      return new Response(
-        (string) $this->t('The Collabora Online editor/viewer is not available.'),
-        Response::HTTP_BAD_REQUEST,
-        ['content-type' => 'text/plain'],
-      );
+      $this->logger->warning('Collabora Online is not available.');
+      throw new BadRequestHttpException('The Collabora Online editor/viewer is not available.');
     }
 
     $current_request_scheme = $request->getScheme();
-    if (!str_starts_with($wopi_client_url, $current_request_scheme . '://')) {
-      $this->logger->error($this->t(
+    if (parse_url($wopi_client_url, PHP_URL_SCHEME) !== $current_request_scheme) {
+      $this->logger->error(
         "The current request uses '@current_request_scheme' url scheme, but the Collabora client url is '@wopi_client_url'.",
         [
           '@current_request_scheme' => $current_request_scheme,
           '@wopi_client_url' => $wopi_client_url,
         ],
-      ));
-      return new Response(
-        (string) $this->t('Viewer error: Protocol mismatch.'),
-        Response::HTTP_BAD_REQUEST,
-        ['content-type' => 'text/plain'],
       );
+      throw new BadRequestHttpException('Viewer error: Protocol mismatch.');
     }
 
     try {
       $render_array = $this->getViewerRender($media, $wopi_client_url, $edit);
     }
     catch (CollaboraNotAvailableException $e) {
-      $this->logger->warning(
-        "Cannot show the viewer/editor.<br>\n" . Error::DEFAULT_ERROR_MESSAGE,
-        Error::decodeException($e) + [],
-      );
-      return new Response(
-        (string) $this->t('The Collabora Online editor/viewer is not available.'),
-        Response::HTTP_BAD_REQUEST,
-        ['content-type' => 'text/plain'],
-      );
+      $this->logger->warning('Cannot show the viewer/editor.');
+      throw new BadRequestHttpException('The Collabora Online editor/viewer is not available.');
```

Comment:
Again, whenever we have a `CollaboraNotAvailableException`, there is something wrong on our side, so even valid requests go wrong. So BAD_REQUEST is wrong.

Reply:
And again we are losing information from the exception message.

Btw, with these regular html routes, the message in a *HttpExceptions is not exposed in the response. At least for 404 and 403. We can test this for 400.

Also, core already logs these exceptions. So maybe we don't need to call the logger ourselves.
We could try what happens if we `throw new ***HttpException('The Collabora Online editor/viewer is not available.', previous: $e)`, so the previous exception is not lost. We would need to check if it is logged in that case.

Reply:
We decided that, for now, we keep the 400 Bad request, even if that would be semantically wrong.
Perhaps we get a better idea in the future.